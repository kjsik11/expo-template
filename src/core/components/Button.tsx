import React from "react";
import {
  useTheme,
  Button as ButtonRNE,
  ButtonProps as ButtonPropsRNE,
} from "@rneui/themed";

export const OUTLINE_BUTTON_BACKGROUND_COLOR_DEFAULT = "rgba(0,0,0,0)";

interface ButtonProps extends ButtonPropsRNE {}

function Button(props: ButtonProps) {
  const { theme } = useTheme();

  const buttonStyle = {
    backgroundColor:
      props.type === "outline"
        ? OUTLINE_BUTTON_BACKGROUND_COLOR_DEFAULT
        : theme.colors.primary,
    borderRadius: 16,
    borderColor: theme.colors.secondary,
    borderWidth: props.type === "outline" ? 0.8 : 0,
  };

  const titleStyle = {
    fontFamily: "Poppins-SemiBold",
    color:
      props.type === "outline" ? theme.colors.secondary : theme.colors.black,
  };

  return (
    <ButtonRNE
      buttonStyle={buttonStyle}
      titleStyle={titleStyle}
      uppercase
      {...props}
    />
  );
}

export default Button;
