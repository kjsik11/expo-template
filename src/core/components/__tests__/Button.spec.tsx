import React from "react";
import {
  render,
  screen,
  fireEvent,
  waitFor,
} from "@testing-library/react-native";
import { ThemeProvider } from "@rneui/themed";
import { ReactTestInstance } from "react-test-renderer";

import THEME from "core/constants/Theme";
import Button, {
  OUTLINE_BUTTON_BACKGROUND_COLOR_DEFAULT,
} from "core/components/Button";

const TITLE_OF_BUTTON_WITH_PRIMARY_TYPE = "Button Primary";
const TEST_ID_OF_BUTTON_WITH_PRIMARY_TYPE = "BUTTON_PRIMARY";

const TITLE_OF_BUTTON_WITH_OUTLINE_TYPE = "Button Outline";
const TEST_ID_OF_BUTTON_WITH_OUTLINE_TYPE = "BUTTON_OUTLINE";

const BUTTON_CALLBACK_FUNCTION = jest.fn(() => null);

describe("ü´Ä Core | üß© Components | üñ≤Ô∏è <Button>", function () {
  let buttonPrimary: ReactTestInstance | null = null;
  let buttonOutline: ReactTestInstance | null = null;

  describe("| Solid Button", function () {
    beforeEach(function () {
      render(
        <ThemeProvider theme={THEME}>
          <Button
            title={TITLE_OF_BUTTON_WITH_PRIMARY_TYPE}
            onPress={BUTTON_CALLBACK_FUNCTION}
            testID={TEST_ID_OF_BUTTON_WITH_PRIMARY_TYPE}
          />
        </ThemeProvider>
      );

      buttonPrimary = screen.getByTestId(TEST_ID_OF_BUTTON_WITH_PRIMARY_TYPE);
    });

    it("should render successfully", () => {
      expect(buttonPrimary).toBeDefined();
    });

    it("should render correctly", () => {
      const buttonAsJson = screen.toJSON();
      expect(buttonAsJson).toMatchSnapshot();
    });

    it("should render with title", () => {
      expect(buttonPrimary).toBeTruthy();
    });

    it("should fires the assigned event on press", async () => {
      fireEvent.press(buttonPrimary as ReactTestInstance);
      fireEvent.press(buttonPrimary as ReactTestInstance);
      await waitFor(() =>
        expect(BUTTON_CALLBACK_FUNCTION.mock.calls.length).toBe(2)
      );
    });
  });

  describe("| Outline Button", function () {
    beforeEach(function () {
      render(
        <ThemeProvider theme={THEME}>
          <Button
            title={TITLE_OF_BUTTON_WITH_OUTLINE_TYPE}
            onPress={BUTTON_CALLBACK_FUNCTION}
            testID={TEST_ID_OF_BUTTON_WITH_OUTLINE_TYPE}
            type="outline"
          />
        </ThemeProvider>
      );

      buttonOutline = screen.getByTestId(TEST_ID_OF_BUTTON_WITH_OUTLINE_TYPE);
    });

    it("should render successfully", () => {
      expect(buttonOutline).toBeDefined();
    });

    it("should render correctly", () => {
      const buttonAsJson = screen.toJSON();
      expect(buttonAsJson).toMatchSnapshot();
    });

    it("should render the background with outline button styles", () => {
      const buttonContainer = (buttonOutline as ReactTestInstance).children[0];
      const backgroundColor =
        buttonContainer &&
        typeof buttonContainer !== "string" &&
        buttonContainer.props.style.backgroundColor;

      expect(backgroundColor).toBe(OUTLINE_BUTTON_BACKGROUND_COLOR_DEFAULT);
    });
  });
});
