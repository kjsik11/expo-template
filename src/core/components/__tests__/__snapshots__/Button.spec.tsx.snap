// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`🫀 Core | 🧩 Components | 🖲️ <Button> | Outline Button should render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "borderRadius": 2,
      },
      undefined,
      false,
    ]
  }
  testID="RNE_BUTTON_WRAPPER"
>
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "busy": false,
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="BUTTON_OUTLINE"
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(0,0,0,0)",
          "borderColor": "#ad1457",
          "borderRadius": 16,
          "borderWidth": 0.8,
          "flexDirection": "row",
          "justifyContent": "center",
          "padding": 8,
          "paddingHorizontal": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#ad1457",
            "fontFamily": "Poppins-SemiBold",
            "fontSize": 18,
            "paddingVertical": 1,
            "textAlign": "center",
            "textTransform": "uppercase",
          }
        }
      >
        Button Outline
      </Text>
    </View>
  </View>
</View>
`;

exports[`🫀 Core | 🧩 Components | 🖲️ <Button> | Solid Button should render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "borderRadius": 2,
      },
      undefined,
      false,
    ]
  }
  testID="RNE_BUTTON_WRAPPER"
>
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "busy": false,
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="BUTTON_PRIMARY"
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "red",
          "borderColor": "#ad1457",
          "borderRadius": 16,
          "borderWidth": 0,
          "flexDirection": "row",
          "justifyContent": "center",
          "padding": 8,
          "paddingHorizontal": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#242424",
            "fontFamily": "Poppins-SemiBold",
            "fontSize": 18,
            "paddingVertical": 1,
            "textAlign": "center",
            "textTransform": "uppercase",
          }
        }
      >
        Button Primary
      </Text>
    </View>
  </View>
</View>
`;
